cpp/src/arrow/compute/exec/sink_node.cc

```diff
@@ -274,11 +284,15 @@ struct OrderBySinkNode final : public SinkNode {
                                               sink_options.backpressure);
   }
 
-  void InputReceived(ExecNode* input, ExecBatch batch) override {
+  void InputReceived(ExecNode* input, std::function<Result<ExecBatch>()> task) override {
     DCHECK_EQ(input, inputs_[0]);
-
-    auto maybe_batch = batch.ToRecordBatch(inputs_[0]->output_schema(),
-                                           plan()->exec_context()->memory_pool());
+    auto batch = task();
+    if (!batch.ok()) {
+      ErrorIfNotOk(batch.status());
+      return;
+    }
+    auto maybe_batch = batch.ValueUnsafe().ToRecordBatch(
+        inputs_[0]->output_schema(), plan()->exec_context()->memory_pool());
```

Comment:
This would build up a task and pass to impl_->InputReceived.  Or maybe the task itself would call impl_->InputReceived and this method calls submit task.

Reply:
I think this should call to SubmitTask, but once again I think this part will need to changes in the next PRs related to the scheduler